File: HomeWork8/8.2/8.2/Tree.cpp
#include "Tree.h"
#include <stdio.h>

struct TreeElement
{
	char function;
	int value;
	TreeElement* parent = nullptr;
	TreeElement* rightChild = nullptr;
	TreeElement* leftChild = nullptr;
};

struct Tree
{
	TreeElement* root = nullptr;
};

Tree* createTree()
{
	return new Tree;
}

int createNumber(char string[], int *counter)
{
	int value = 0;
	while (string[*counter] >= '0' && string[*counter] <= '9')
	{
		value = value * 10 + (string[*counter] - '0');
		(*counter)++;
	}
	return value;
}

TreeElement* addElementInTree(char string[], int *counter, TreeElement* parent)
{
	if (string[*counter] == '\0')
	{
		return nullptr;
	}
	while (string[*counter] == '(' || string[*counter] == ')' || string[*counter] == ' ')
	{
		(*counter)++;
		if (string[*counter] == '\0')
		{
			return nullptr;
		}
	}
	if (string[*counter] >= '0' && string[*counter] <= '9')
	{
		const int value = createNumber(string, counter);
		return new TreeElement{ ' ', value, parent, nullptr, nullptr };
	}
	else
	{
		char function = string[*counter];
		(*counter)++;
		auto newElement = new TreeElement{ function, 0, parent, nullptr, nullptr };
		newElement->leftChild = addElementInTree(string, counter, newElement);
		newElement->rightChild = addElementInTree(string, counter, newElement);
		return newElement;
	}
}

void buildTree(Tree* tree, char string[])
{
	int counter = 0;
	tree->root = addElementInTree(string, &counter, tree->root);
}

void outputTreeElement(TreeElement* element)
{
	if (element == nullptr)
	{
		return;
	}
	if (element->function == ' ')
	{
		printf("%d ", element->value);
	}
	else
	{
		printf("%c ", element->function);
	}
	outputTreeElement(element->leftChild);
	outputTreeElement(element->rightChild);
}

void outputTree(Tree* tree)
{
	outputTreeElement(tree->root);
}

TreeElement* counting(TreeElement* element)
{
	if (element->function != ' ')
	{
		TreeElement* leftElement = counting(element->leftChild);
		TreeElement* rightElement = counting(element->rightChild);
		if (element->function == '+')
		{
			element->function = ' ';
			element->value = leftElement->value + rightElement->value;
		}
		if (element->function == '-')
		{
			element->function = ' ';
			element->value = leftElement->value - rightElement->value;
		}
		if (element->function == '*')
		{
			element->function = ' ';
			element->value = leftElement->value * rightElement->value;
		}
		if (element->function == '/')
		{
			element->function = ' ';
			element->value = leftElement->value / rightElement->value;
		}
	}
	return element;
}

void toCountResult(Tree* tree)
{
	counting(tree->root);
}

void deleteTreeElement(TreeElement* element)
{
	if (element->leftChild == nullptr && element->rightChild == nullptr)
	{
		delete element;
	}
	else
	{
		if (element->leftChild != nullptr)
		{
			deleteTreeElement(element->leftChild);
		}
		if (element->rightChild != nullptr)
		{
			deleteTreeElement(element->rightChild);
		}
		delete element;
	}
}
int getValueOfRoot(Tree* tree)
{
	return tree->root->value;
}

void deleteTree(Tree* tree)
{
	deleteTreeElement(tree->root);
	delete tree;
}

File: 2Sem.HomeWork4/2.4.1/2.4.1/Tree.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Text;

namespace _2._4._1
{
	/// <summary>
	/// Tree's class
	/// </summary>
	public class Tree
    {
		private INode root;

		/// <summary>
		/// Check expression for correct
		/// </summary>
		/// <param name="expression">String with expression</param>
		/// <returns>Correct of not</returns>
		public bool IsExpressionCorrect(string expression)
		{
			int amountOfValues = 0;

			for (int i = 0; i < expression.Length; i++)
			{
				if (expression[i] == ' ' || expression[i] == ')' || expression[i] == '(')
				{
					continue;
				}
				else if (char.IsDigit(expression[i]))
				{
					CreateNumber(ref i, expression);
					amountOfValues++;
					continue;
				}
				else if (expression[i] == '+' || expression[i] == '-' || expression[i] == '*' || expression[i] == '/')
				{
					amountOfValues--;
				}
				else
				{
					return false;
				}
			}

			return amountOfValues != 0;
		}

		/// <summary>
		/// Create the tree by expression
		/// </summary>
		/// <param name="str">String with expression</param>
		public void BuildTree(string str)
		{
			if (!IsExpressionCorrect(str))
			{
				throw new InvalidExpressionException();
			}
			int counter = 0;
			root = AddElementInTree(str, ref counter);
		}

		/// <summary>
		/// Get operation for counting
		/// </summary>
		/// <param name="operation">Addition, division, multiplication or subtraction</param>
		/// <returns>Operation</returns>
		private Operation ChoiceOperation(char operation)
		{
			Operation newElement;
			switch (operation)
			{
				case '+':
					newElement = new Addition();
					newElement.OperationSign = '+';
					break;

				case '-':
					newElement = new Subtraction();
					newElement.OperationSign = '-';
					break;

				case '*':
					newElement = new Multiplication();
					newElement.OperationSign = '*';
					break;

				default:
					newElement = new Division();
					newElement.OperationSign = '/';
					break;
			}
			return newElement;
		}

		/// <summary>
		/// Add element in tree
		/// </summary>
		/// <param name="str">String with expression</param>
		/// <param name="counter">Counter of string</param>
		/// <returns>New tree element</returns>
		private INode AddElementInTree(string str, ref int counter)
		{
			if (counter == str.Length)
			{
				return null;
			}

			while (str[counter] == '(' || str[counter] == ')' || str[counter] == ' ')
			{
				counter++;
				if (counter == str.Length)
				{
					return null;
				}
			}
			
			if (char.IsDigit(str[counter]))
			{
				int value = CreateNumber(ref counter, str);
				return new Number(value);
			}

			else
			{
				char operation = str[counter];
				counter++;
				var newElement = ChoiceOperation(operation);

				newElement.LeftChild = AddElementInTree(str, ref counter);
				newElement.RightChild = AddElementInTree(str, ref counter);
				return newElement;
			}
		}

		/// <summary>
		/// Get number from string
		/// </summary>
		/// <param name="position">Counter of string</param>
		/// <param name="expression">String of expression</param>
		/// <returns>Number</returns>
		private static int CreateNumber(ref int position, string expression)
		{
			string value = null;

			while (char.IsDigit(expression[position]))
			{
				value += expression[position];
				position++;
			}

			return int.Parse(value);
		}

		/// <summary>
		/// Counting expression
		/// </summary>
		/// <returns>Result of expression</returns>
		public double Counting() => root.Counting();

		/// <summary>
		/// Output exxpression
		/// </summary>
		public void OutputTree(){ root.Print(); }
	}
}
